### PR TITLE
Fix garbage-collect --delete-untagged to handle schema 2 manifest list and OCI image index

### LINKED ISSUE #3178: Registry garbage-collect --delete-untagged removes multi-arch manifests
Hello!
We are currently experimenting with using multi-arch images in our infrastructure, as we have a mix of IBM Power and regular AMD64-based machines, and are currently using different tags/repositories to distinguish between the architectures, which is not ideal. 

We found that the garbage-collect command with the -m parameter on the registry removes the manifests that enable us to use multi-arch, thus rendering us unable to run garbage collection and keep our registry somewhat clean and space-efficient. 

I have compiled the following script which shows the behaviour. Beware that it runs `docker system prune --all --volumes`:
```
# Remove old registry, clean slate
docker rm -f registry
docker system prune --all --volumes -f

# Start registry
docker run -d -p 5000:5000 --name registry registry

# Pull AMD64-image
docker pull "docker"
docker tag "docker" "localhost:5000/amd64/docker:latest"
docker push "localhost:5000/amd64/docker:latest"

# ARM-image
docker pull "ppc64le/docker"
docker tag "ppc64le/docker" "localhost:5000/ppc64le/docker:latest"
docker push "localhost:5000/ppc64le/docker:latest"

#Create and push manifest for multi-arch ubuntu
docker manifest create --insecure "localhost:5000/docker:latest" "localhost:5000/amd64/docker:latest" "localhost:5000/ppc64le/docker:latest"
docker manifest push -p "localhost:5000/docker:latest"

# Clear local cache
docker system prune --all -f

# Garbage-collect registry
docker exec registry registry garbage-collect --delete-untagged /etc/docker/registry/config.yml

# Pull multi-arch image
docker pull "localhost:5000/docker:latest"
```
The last couple of lines of output on my machine is: 
```
Total reclaimed space: 439.4MB
amd64/docker
amd64/docker: marking manifest sha256:5782329618eaf703204c70f61c1697a428275bc8d8e5a58fa324798e78e31937 
amd64/docker: marking blob sha256:61d75b305666b5c672c2dc789c526f405402c3d37eb44a77470e9217790b9a41
amd64/docker: marking blob sha256:530afca65e2ea04227630ae746e0c85b2bd1a179379cbf2b6501b49c4cab2ccc
amd64/docker: marking blob sha256:33702c1843d19cf7c37af730bcf50c456ac6456ed789053432b000db75d3bed3
amd64/docker: marking blob sha256:d1c203384d5b9b22606055f5a6708153653df701158a6c04748cb77be8238c9e
amd64/docker: marking blob sha256:146feb07c33136aba6d87c2a8d6882cd4d438d957eaaa8f388f59214f1269bd0
amd64/docker: marking blob sha256:cee6b871713b8386f334ada3f80a6f187b8e9130ce7d69236e34fee9f9d44556
amd64/docker: marking blob sha256:db8946a7c6c2c651e2344e7e9ebb692ff53ae6ab2b108a5f484384e165056beb
amd64/docker: marking blob sha256:649b2db28c4901ee36039690e8086767f9c069ebef2f819f53937774a4b74d6b
amd64/docker: marking blob sha256:e7b044ff4e6b0f04ba45ff9f486263b0741ce55315ac6fb0498aec49a0375ee0
amd64/docker: marking blob sha256:dbd7811a8fce869b1e3e2698db341271fcedfe91a3c90e63b19b243ba4187d91
docker
manifest eligible for deletion: sha256:18d3c182508d401451fe7a0c538dff3686360456b233675b4ddb2371aa9d1608
docker: marking manifest sha256:1abd08c6dc1cea1308a4f3213cdb7d7264548ab21beedbde30aa7eed7f209d2d 
docker: marking blob sha256:5782329618eaf703204c70f61c1697a428275bc8d8e5a58fa324798e78e31937
docker: marking blob sha256:18d3c182508d401451fe7a0c538dff3686360456b233675b4ddb2371aa9d1608
manifest eligible for deletion: sha256:5782329618eaf703204c70f61c1697a428275bc8d8e5a58fa324798e78e31937
ppc64le/docker
ppc64le/docker: marking manifest sha256:18d3c182508d401451fe7a0c538dff3686360456b233675b4ddb2371aa9d1608 
ppc64le/docker: marking blob sha256:a10903a7995b02bb2bb17f3be3423e93bd2ff6a70618d0d42cfed01639a17ec9
ppc64le/docker: marking blob sha256:d6201b52ea9b908d4d6e950e2a79ded27be48979d6f0f63bd3a57b16b621f188
ppc64le/docker: marking blob sha256:7c5cedde51de1c8ffccd5521fd02fc1efc1cc44ece2d5dccb1e550a65366cd80
ppc64le/docker: marking blob sha256:7ab4b28968b92237adc6438d06c14d4680a3936996176f2cea20d781885af568
ppc64le/docker: marking blob sha256:387f12b774ebb4a88b15e4f754eb9c772618a15230f9c093759496c41f5c6c8f
ppc64le/docker: marking blob sha256:8847e0f1fbcbc029817eb1d930620fd68a65b6bf9dcb73b6d2e96d4e8161ba75
ppc64le/docker: marking blob sha256:b7425929e73f782470c79b835b77e1d923e63be874374d247c1deb271178256f
ppc64le/docker: marking blob sha256:72d865172b780f7e0c045e1d59db4d57be4d6f52d93a27f7d7444edead40260b
time="2022-08-01T13:30:53.876986875Z" level=info msg="deleting manifest: /docker/registry/v2/repositories/docker/_manifests/revisions/sha256/18d3c182508d401451fe7a0c538dff3686360456b233675b4ddb2371aa9d1608" go.version=go1.16.15 instance.id=f87bf88e-d8d2-4dfa-9095-71d90367e1aa service=registry 
time="2022-08-01T13:30:53.877150402Z" level=info msg="deleting manifest: /docker/registry/v2/repositories/docker/_manifests/revisions/sha256/5782329618eaf703204c70f61c1697a428275bc8d8e5a58fa324798e78e31937" go.version=go1.16.15 instance.id=f87bf88e-d8d2-4dfa-9095-71d90367e1aa service=registry 

21 blobs marked, 0 blobs and 2 manifests eligible for deletion
latest: Pulling from docker
manifest for localhost:5000/docker:latest not found: manifest unknown: manifest unknown

```

I have confirmed that the issue is present on my workstation, which is running:
docker version:
```
Client: Docker Engine - Community
 Version:           20.10.17
 API version:       1.41
 Go version:        go1.17.11
 Git commit:        100c701
 Built:             Mon Jun  6 23:02:57 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.11
  Git commit:       a89b842
  Built:            Mon Jun  6 23:01:03 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.6
  GitCommit:        10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc:
  Version:          1.1.2
  GitCommit:        v1.1.2-0-ga916309
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```
docker info:
```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 5
  Running: 5
  Paused: 0
  Stopped: 0
 Images: 27
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc version: v1.1.2-0-ga916309
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.15.0-41-generic
 Operating System: Ubuntu 22.04 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 31.19GiB
 Name: FLPC8286
 ID: OF2W:RJAA:ZPBP:WO6M:H4GP:GDZC:R5PE:PARX:O3CO:2BPW:CM4U:KJHB
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```
docker exec registry registry --version:
```
registry github.com/docker/distribution v2.8.1+unknown
```

### PR DESCRIPTION
**Pull request by thewolt**
resolves #3178

### CODE REVIEW

**Comment by milosgajdos** [comment]
This is failing the linter and needs fixing

**Comment by thewolt** [comment]
@milosgajdos the issues reported by the linter should be fixed

**Comment by thewolt** [comment]
> How is this pull request coming along? If some help is needed I'd be willing to contribute to this issue - just let me know.

I don't understand what else is needed? all the checks have passed. Do I need to rebase onto origin/main ?

**Comment by WhySoBad** [comment]
> I don't understand what else is needed? all the checks have passed

The question was more directed at @milosgajdos since he seems to be the one assigned with this issue. 
For me the issue seems completed but I don't know if the folks at docker are missing something. 
> Do I need to rebase onto origin/main ?

I think a rebase onto origin/main can't do any harm since it's necessary to rebase before merging anyways. 

Sorry if my previous comment sounded like I was trying to stress you out. I really appreciate your work on this issue!

**Comment by milosgajdos** [comment]
> The question was more directed at @milosgajdos since he seems to be the one assigned with this issue.

No, I am NOT assigned this issue. This is all done in everyone's free unpaid time. I merely mentioned the CI failure in my comment.

**Comment by thewolt** [comment]
@milosgajdos @WhySoBad so can anyone point me to two reviewers ?

**Comment by godber** [comment]
@thewolt We have no authority to merge but we reviewed and used your patch against 3.0.0-alpha and rolled it out to prod because it looked good, had tests that passed, worked in our additional testing and at that point any risk that posed was far less than the constant problems posed by a destructive garbage collection mechanism.

It looks like milosgajdos can merge but I've looked through the merged PRs and he's the only one who has merged since February 29th when wy65701436 merged https://github.com/distribution/distribution/pull/4286.  So sadly it seems he's been going it alone recently.

There's a [MAINTAINERS File](https://github.com/distribution/distribution/blob/main/MAINTAINERS) and a [GOVERNANCE](https://github.com/distribution/distribution/blob/main/GOVERNANCE.md) document if you want to dig further or solicit more help somehow.

**Comment by WhySoBad** [comment]
> No, I am NOT assigned this issue. This is all done in everyone's free unpaid time. I merely mentioned the CI failure in my comment.

I apologize for my faulty assumption. Based on your activity in this project and your github profile it seemed as you were employed by docker.

**Comment by milosgajdos** [comment]
> I apologize for my faulty assumption. Based on your activity in this project and your github profile it seemed as you were employed by docker.

I _used_ to be employed by Docker inc. but they never paid me to work on this project. Assuming someone who works for Docker automatically qualifies as a person who works on this project is a false assumption.

This project was _donated_ to CNCF by Docker a few years ago. All the work I've done on this project has been in my free and unpaid time.

No hurt feelings, I just need to clarify all the wrong assumptions made here for the future generations 😄

**Comment by godber** [comment]
There's some bottleneck in the CNCF process around review and merge.  The sad irony is I DO pay people to work on CNCF projects ... most of our issues and PRs have gone ignored and closed by bots.

I guess I should figure out how to get more involved in the CNCF but I've assumed since we don't work for some unicorn we'll be ignored.  Especially given recent supply chain hysteria.  But I digress, this isn't the venue for this discussion.

File: registry/storage/garbagecollect.go [hunk 1]
```diff
@@ -84,18 +86,14 @@ func MarkAndSweep(ctx context.Context, storageDriver driver.StorageDriver, regis
...
-
-			descriptors := manifest.References()
-			for _, descriptor := range descriptors {
-				markSet[descriptor.Digest] = struct{}{}
-				emit("%s: marking blob %s", repoName, descriptor.Digest)
-			}
-
-			return nil
+			return markManifestReferences(dgst, manifestService, ctx, func(d digest.Digest) bool {
+				_, exists := markSet[d]
```
[the remark this answers is not in the document]
Comment (thewolt) [reply registry/storage/garbagecollect.go]: you're right... renamed the var to `marked`

**Review by milosgajdos — APPROVED** [review]
LGTM, @Jamstah mind having a look?

**Comment by nmapx** [comment]
Can someone double check this PR and merge it finally? I feel this fix is somehow underrated unfortunately. There are plenty of people waiting for it but I guess none of them are allowed to click the merge button. So frustrating.

**Comment by milosgajdos** [comment]
PTAL @thaJeztah @squizzi

**Comment by gislikonrad** [comment]
Yeah, this bug caused a huge issue for us after we started building multiplatform containers. First garbage-collect after that was catastrophic. Please merge.

File: registry/storage/garbagecollect_test.go [hunk 2]
```diff
@@ -479,3 +625,225 @@ func TestOrphanBlobDeleted(t *testing.T) {
...
+	})
+	if err != nil {
+		t.Fatalf("Failed mark and sweep: %v", err)
+	}
+
+	after := allBlobs(t, registry)
+	afterManifests := allManifests(t, manifestService)
+
+	if _, ok := after[dgst]; ok {
+		t.Fatalf("Untaged manifestlist still exists")
```
Comment (squizzi) [inline registry/storage/garbagecollect_test.go:786]:
```suggestion
		t.Fatalf("Untagged manifestlist still exists")
```

File: registry/storage/garbagecollect_test.go [hunk 3]
```diff
@@ -479,3 +625,225 @@ func TestOrphanBlobDeleted(t *testing.T) {
...
+	if _, ok := afterManifests[image1.manifestDigest]; !ok {
+		t.Fatalf("First manifest is missing")
+	}
+
+	if _, ok := afterManifests[image2.manifestDigest]; !ok {
+		t.Fatalf("Second manifest is missing")
+	}
+}
+
+func TestTaggedManifestlistWithDeletedreference(t *testing.T) {
```
Comment (squizzi) [inline registry/storage/garbagecollect_test.go:798]:
```suggestion
func TestTaggedManifestlistWithDeletedReference(t *testing.T) {
```

**Review by squizzi — APPROVED** [review]

**Comment by thewolt** [comment]
@squizzi Thanks for the review, i've updated the test case commit with your suggestions

**Comment by milosgajdos** [comment]
@thewolt can you squash, please.

**Comment by thewolt** [comment]
> @thewolt can you squash, please.

Squash done 👍

**Comment by nmapx** [comment]
@milosgajdos when is it gonna be released?

**Comment by milosgajdos** [comment]
> @milosgajdos when is it gonna be released?

Short answer: there is no concrete plan

Long answer: this project is run by unpaid volunteers who have their own full-time jobs, so it all depends on their time availability TBD

File: registry/storage/garbagecollect.go [hunk 4]
```diff
@@ -68,12 +68,14 @@ func MarkAndSweep(ctx context.Context, storageDriver driver.StorageDriver, regis
 					return fmt.Errorf("failed to retrieve tags for digest %v: %v", dgst, err)
 				}
 				if len(tags) == 0 {
-					emit("manifest eligible for deletion: %s", dgst)
 					// fetch all tags from repository
 					// all of these tags could contain manifest in history
 					// which means that we need check (and delete) those references when deleting manifest
 					allTags, err := repository.Tags(ctx).All(ctx)
 					if err != nil {
+						if _, ok := err.(distribution.ErrManifestUnknownRevision); !ok {
```
Comment (microyahoo) [inline registry/storage/garbagecollect.go:76]: hi @thewolt, @squizzi, I have a question about `ErrManifestUnknownRevision`, under what circumstances would this error occur? According to the [code](https://github.com/distribution/distribution/blob/d61d8ebc16770c0b257cb342f12576d59e2df272/registry/storage/tagstore.go#L38), it should return an `ErrRepositoryUnknown` error, please correct me if I have mistaken.

Additionally, I have also submitted a [PR](https://github.com/distribution/distribution/pull/4331) related to this.

Comment (thewolt) [reply registry/storage/garbagecollect.go:76]: Sorry but I've never used Go before this PR.
When I brought the test cases from readerx, I found that adding this condition fixed the test and I looked no further.

Using the debugger, it's clear that the error i wanted to catch was indeed `ErrRepositoryUnknown` and the way i've put the condition is useless.

So yes, your PR is correct and provides a more useful message.

Comment (microyahoo) [reply registry/storage/garbagecollect.go:76]: Thanks for your clarification, I will update it.

**Comment by gislikonrad** [comment]
Any update on when a new alpha of 3.0.0 will released so that we can use this fix without building the container ourselves?

**Pull request merged**